File: machine/cpp/machine/runtime/frame/loop.cpp
#include "machine/runtime/frame/loop.h"
#include "machine/runtime/cpu_executor.h"
#include "machine/runtime/runtime.h"
#include "machine/scheduler/device.h"

#include <limits>

namespace bmsx {
void FrameLoopState::reset() {
	frameState = FrameState{};
	frameActive = false;
}

FrameLoopStateSnapshot FrameLoopState::captureState() const {
	return FrameLoopStateSnapshot{frameState, frameActive};
}

void FrameLoopState::restoreState(const FrameLoopStateSnapshot& state) {
	reset();
	frameState = state.frameState;
	frameActive = state.frameActive;
}

void FrameLoopState::resetFrameState(Runtime& runtime) {
	frameActive = false;
	runtime.vblank.abandonTick();
	runtime.machine.cpu.clearHaltUntilIrq();
	runtime.frameScheduler.reset();
	reset();
	runtime.frameScheduler.resetTickTelemetry();
}

void FrameLoopState::beginFrameState(Runtime& runtime, i64 budget, i64 carry) {
	frameActive = true;
	runtime.vblank.beginTick();
	frameState = FrameState{};
	frameState.cycleBudgetRemaining = budget;
	frameState.cycleBudgetGranted = budget;
	frameState.cycleCarryGranted = carry;
}

void FrameLoopState::abandonFrameState(Runtime& runtime) {
	frameActive = false;
	runtime.vblank.abandonTick();
}

bool FrameLoopState::consumeSystemReset(Runtime& runtime) {
	if (!runtime.machine.systemController.takeResetRequest()) {
		return false;
	}
	runtime.rebootSystem();
	return true;
}

bool FrameLoopState::prepareScheduledFrame(Runtime& runtime) {
	if (!frameActive) {
		return runtime.frameScheduler.startScheduledFrame(runtime);
	}
	if (frameState.cycleBudgetRemaining <= 0) {
		return runtime.frameScheduler.refillFrameBudget(runtime, frameState);
	}
	return true;
}

void FrameLoopState::finalizeUpdateSlice(Runtime& runtime) {
	if (runtime.m_pendingCall == Runtime::PendingCall::Entry && !runtime.vblank.tickCompleted()) {
		return;
	}
	abandonFrameState(runtime);
}

void FrameLoopState::runActiveFrameState(Runtime& runtime) {
	if (runtime.m_pendingCall == Runtime::PendingCall::Entry) {
		runUpdatePhase(runtime);
		frameState.updateExecuted = runtime.m_pendingCall != Runtime::PendingCall::Entry;
		finalizeUpdateSlice(runtime);
		return;
	}
	finalizeUpdateSlice(runtime);
}

InstructionStepResult FrameLoopState::runActiveFrameInstruction(Runtime& runtime) {
	if (runtime.m_pendingCall != Runtime::PendingCall::Entry) {
		finalizeUpdateSlice(runtime);
		return InstructionStepResult::Advanced;
	}
	const InstructionStepResult result = runUpdateInstruction(runtime);
	frameState.updateExecuted = runtime.m_pendingCall != Runtime::PendingCall::Entry;
	finalizeUpdateSlice(runtime);
	return result;
}

void FrameLoopState::runUpdatePhase(Runtime& runtime) {
	auto& cpu = runtime.machine.cpu;
	while (true) {
		if (cpu.isHaltedUntilIrq() || runtime.machine.systemController.cpuHeld()) {
			const bool tickCompleted = runtime.cpuExecution.runStoppedCpu(runtime, frameState);
			if (tickCompleted || cpu.isHaltedUntilIrq() || runtime.machine.systemController.cpuHeld()) {
				return;
			}
			continue;
		}
		if (runtime.m_pendingCall != Runtime::PendingCall::Entry) {
			return;
		}
		const RunResult result = runtime.cpuExecution.runWithBudget(runtime, frameState);
		if (consumeSystemReset(runtime)) {
			return;
		}
		if (result == RunResult::Halted && cpu.getFrameDepth() == 0) {
			runtime.m_pendingCall = Runtime::PendingCall::None;
			runtime.frameScheduler.clearQueuedTime();
			abandonFrameState(runtime);
			return;
		}
		if (cpu.isHaltedUntilIrq()) {
			return;
		}
		return;
	}
}

InstructionStepResult FrameLoopState::runUpdateInstruction(Runtime& runtime) {
	auto& cpu = runtime.machine.cpu;
	if (cpu.isHaltedUntilIrq() || runtime.machine.systemController.cpuHeld()) {
		const i64 previousRemaining = frameState.cycleBudgetRemaining;
		const bool tickCompleted = runtime.cpuExecution.runStoppedCpu(runtime, frameState);
		if (tickCompleted || cpu.isHaltedUntilIrq() || runtime.machine.systemController.cpuHeld()) {
			return tickCompleted || frameState.cycleBudgetRemaining != previousRemaining
				? InstructionStepResult::Advanced
				: InstructionStepResult::Blocked;
		}
	}
	if (runtime.m_pendingCall != Runtime::PendingCall::Entry) {
		return InstructionStepResult::Blocked;
	}
	const InstructionStepResult result = runtime.cpuExecution.runInstruction(runtime, frameState);
	if (cpu.getFrameDepth() == 0) {
		runtime.m_pendingCall = Runtime::PendingCall::None;
		runtime.frameScheduler.clearQueuedTime();
		abandonFrameState(runtime);
	}
	return result;
}

bool FrameLoopState::tickUpdate(Runtime& runtime) {
	using PendingCall = Runtime::PendingCall;
	if (consumeSystemReset(runtime)) {
		return true;
	}

	const bool previousFrameActive = frameActive;
	const i64 previousRemaining = previousFrameActive ? frameState.cycleBudgetRemaining : -1;
	const bool previousPending = runtime.m_pendingCall == PendingCall::Entry;
	const i64 previousSequence = runtime.frameScheduler.lastTickSequence;
	if (!prepareScheduledFrame(runtime)) {
		return false;
	}

	runActiveFrameState(runtime);
	if (frameActive
		&& (runtime.machine.cpu.isHaltedUntilIrq() || runtime.machine.systemController.cpuHeld())
		&& runtime.machine.scheduler.nextDeadline() == std::numeric_limits<i64>::max()) {
		// Cart parked waiting for an interrupt that nothing has scheduled: report
		// no progress so the scheduler yields the host slice instead of spinning.
		return false;
	}
	const bool nextFrameActive = frameActive;
	if (nextFrameActive != previousFrameActive) {
		return true;
	}
	if (nextFrameActive && frameState.cycleBudgetRemaining != previousRemaining) {
		return true;
	}
	if ((runtime.m_pendingCall == PendingCall::Entry) != previousPending) {
		return true;
	}
	return runtime.frameScheduler.lastTickSequence != previousSequence;
}

InstructionStepResult FrameLoopState::tickInstruction(Runtime& runtime) {
	using PendingCall = Runtime::PendingCall;
	if (consumeSystemReset(runtime)) {
		return InstructionStepResult::Advanced;
	}
	const bool previousFrameActive = frameActive;
	const i64 previousRemaining = previousFrameActive ? frameState.cycleBudgetRemaining : -1;
	const bool previousPending = runtime.m_pendingCall == PendingCall::Entry;
	const i64 previousSequence = runtime.frameScheduler.lastTickSequence;
	if (!prepareScheduledFrame(runtime)) {
		return InstructionStepResult::Blocked;
	}
	const InstructionStepResult result = runActiveFrameInstruction(runtime);
	if (result == InstructionStepResult::Executed) {
		return result;
	}
	if (frameActive != previousFrameActive
		|| (frameActive && frameState.cycleBudgetRemaining != previousRemaining)
		|| ((runtime.m_pendingCall == PendingCall::Entry) != previousPending)
		|| runtime.frameScheduler.lastTickSequence != previousSequence) {
		return InstructionStepResult::Advanced;
	}
	return result;
}
} // namespace bmsx

File: machine/ts/machine/runtime/frame/loop.ts
import {
	InstructionStepResult,
	type FrameState,
} from './state';
import { Runtime } from '../runtime';
import { RunResult } from '../../cpu/cpu';

export type FrameLoopStateSnapshot = {
	frameState: FrameState;
	frameActive: boolean;
	frameDeltaMs: number;
};

export class FrameLoopState {
	public currentTimeMs = 0;
	public frameDeltaMs = 0;
	public readonly frameState: FrameState = {
		updateExecuted: false,
		cycleBudgetRemaining: 0,
		cycleBudgetGranted: 0,
		cycleCarryGranted: 0,
		activeCpuUsedCycles: 0,
	};
	public frameActive = false;

	/** The in-flight frame state while a frame is active, or null when idle. */
	public get currentFrameState(): FrameState | null {
		return this.frameActive ? this.frameState : null;
	}

	constructor(private readonly runtime: Runtime) {
	}

	public reset(): void {
		this.currentTimeMs = 0;
		this.frameDeltaMs = 0;
		this.frameActive = false;
		const state = this.frameState;
		state.updateExecuted = false;
		state.cycleBudgetRemaining = 0;
		state.cycleBudgetGranted = 0;
		state.cycleCarryGranted = 0;
		state.activeCpuUsedCycles = 0;
	}

	public captureState(): FrameLoopStateSnapshot {
		return {
			frameState: { ...this.frameState },
			frameActive: this.frameActive,
			frameDeltaMs: this.frameDeltaMs,
		};
	}

	public restoreState(snapshot: FrameLoopStateSnapshot): void {
		this.reset();
		const state = this.frameState;
		state.updateExecuted = snapshot.frameState.updateExecuted;
		state.cycleBudgetRemaining = snapshot.frameState.cycleBudgetRemaining;
		state.cycleBudgetGranted = snapshot.frameState.cycleBudgetGranted;
		state.cycleCarryGranted = snapshot.frameState.cycleCarryGranted;
		state.activeCpuUsedCycles = snapshot.frameState.activeCpuUsedCycles;
		this.frameActive = snapshot.frameActive;
		this.frameDeltaMs = snapshot.frameDeltaMs;
	}

	public resetFrameState(): void {
		const runtime = this.runtime;
		this.abandonFrameState();
		runtime.machine.cpu.clearHaltUntilIrq();
		runtime.frameScheduler.reset();
		this.reset();
		runtime.frameScheduler.resetTickTelemetry();
	}

	public beginFrameState(budget: number, carry: number): FrameState {
		const runtime = this.runtime;
		this.frameDeltaMs = runtime.timing.frameDurationMs;
		const state = this.frameState;
		state.updateExecuted = false;
		state.cycleBudgetRemaining = budget;
		state.cycleBudgetGranted = budget;
		state.cycleCarryGranted = carry;
		state.activeCpuUsedCycles = 0;
		runtime.vblank.beginTick();
		this.frameActive = true;
		return state;
	}

	public tickUpdate(): boolean {
		const runtime = this.runtime;
		if (this.consumeSystemReset()) {
			return true;
		}
		const previousFrameActive = this.frameActive;
		const previousRemaining = previousFrameActive ? this.frameState.cycleBudgetRemaining : -1;
		const frameScheduler = runtime.frameScheduler;
		const previousPendingEntry = runtime.pendingCall === 'entry';
		const previousSequence = frameScheduler.lastTickSequence;
		if (!this.prepareScheduledFrame()) {
			return false;
		}
		this.runActiveFrameState();
		if (this.frameActive
			&& (runtime.machine.cpu.isHaltedUntilIrq() || runtime.machine.systemController.cpuHeld())
			&& runtime.machine.scheduler.nextDeadline() === Number.MAX_SAFE_INTEGER) {
			// Cart parked waiting for an interrupt that nothing has scheduled: report
			// no progress so the scheduler yields the host slice instead of spinning.
			return false;
		}
		const nextFrameActive = this.frameActive;
		if (nextFrameActive !== previousFrameActive) {
			return true;
		}
		if (nextFrameActive && this.frameState.cycleBudgetRemaining !== previousRemaining) {
			return true;
		}
		const nextPendingCall = runtime.pendingCall;
		if ((nextPendingCall === 'entry') !== previousPendingEntry) {
			return true;
		}
		const nextSequence = frameScheduler.lastTickSequence;
		return nextSequence !== previousSequence;
	}

	public tickInstruction(): InstructionStepResult {
		const runtime = this.runtime;
		if (this.consumeSystemReset()) {
			return InstructionStepResult.Advanced;
		}
		const previousFrameActive = this.frameActive;
		const previousRemaining = previousFrameActive ? this.frameState.cycleBudgetRemaining : -1;
		const previousPendingEntry = runtime.pendingCall === 'entry';
		const previousSequence = runtime.frameScheduler.lastTickSequence;
		if (!this.prepareScheduledFrame()) {
			return InstructionStepResult.Blocked;
		}
		const result = this.runActiveFrameInstruction();
		if (result === InstructionStepResult.Executed) {
			return result;
		}
		if (this.frameActive !== previousFrameActive
			|| (this.frameActive && this.frameState.cycleBudgetRemaining !== previousRemaining)
			|| ((runtime.pendingCall === 'entry') !== previousPendingEntry)
			|| runtime.frameScheduler.lastTickSequence !== previousSequence) {
			return InstructionStepResult.Advanced;
		}
		return result;
	}

	public abandonFrameState(): void {
		this.frameActive = false;
		const runtime = this.runtime;
		runtime.vblank.abandonTick();
	}

	private consumeSystemReset(): boolean {
		const runtime = this.runtime;
		if (!runtime.machine.systemController.takeResetRequest()) {
			return false;
		}
		runtime.rebootSystem();
		return true;
	}

	private prepareScheduledFrame(): boolean {
		if (!this.frameActive) {
			return this.runtime.frameScheduler.startScheduledFrame();
		}
		if (this.frameState.cycleBudgetRemaining <= 0) {
			return this.runtime.frameScheduler.refillFrameBudget(this.frameState);
		}
		return true;
	}

	private runActiveFrameState(): void {
		const runtime = this.runtime;
		const state = this.frameState;
		if (runtime.pendingCall === 'entry') {
			this.runUpdatePhase();
			state.updateExecuted = runtime.pendingCall !== 'entry';
			this.finalizeUpdateSlice();
			return;
		}
		this.finalizeUpdateSlice();
	}

	private runActiveFrameInstruction(): InstructionStepResult {
		const runtime = this.runtime;
		if (runtime.pendingCall !== 'entry') {
			this.finalizeUpdateSlice();
			return InstructionStepResult.Advanced;
		}
		const result = this.runUpdateInstruction();
		this.frameState.updateExecuted = runtime.pendingCall !== 'entry';
		this.finalizeUpdateSlice();
		return result;
	}

	private finalizeUpdateSlice(): void {
		const runtime = this.runtime;
		if (runtime.pendingCall === 'entry' && !runtime.vblank.tickCompleted) {
			return;
		}
		this.abandonFrameState();
	}

	private runUpdatePhase(): void {
		const runtime = this.runtime;
		const state = this.frameState;
		const cpu = runtime.machine.cpu;
		const cpuExecution = runtime.cpuExecution;
		while (true) {
			if (cpu.isHaltedUntilIrq() || runtime.machine.systemController.cpuHeld()) {
				const tickCompleted = cpuExecution.runStoppedCpu(state);
				if (tickCompleted || cpu.isHaltedUntilIrq() || runtime.machine.systemController.cpuHeld()) {
					return;
				}
				continue;
			}
			if (runtime.pendingCall !== 'entry') {
				return;
			}
			const result = runtime.cpuExecution.runWithBudget(state);
			if (this.consumeSystemReset()) {
				return;
			}
			if (result === RunResult.Halted && cpu.getFrameDepth() === 0) {
				runtime.pendingCall = null;
				runtime.frameScheduler.clearQueuedTime();
				this.abandonFrameState();
				return;
			}
			if (cpu.isHaltedUntilIrq()) {
				return;
			}
			return;
		}
	}

	private runUpdateInstruction(): InstructionStepResult {
		const runtime = this.runtime;
		const state = this.frameState;
		const cpu = runtime.machine.cpu;
		if (cpu.isHaltedUntilIrq() || runtime.machine.systemController.cpuHeld()) {
			const previousRemaining = state.cycleBudgetRemaining;
			const tickCompleted = runtime.cpuExecution.runStoppedCpu(state);
			if (tickCompleted || cpu.isHaltedUntilIrq() || runtime.machine.systemController.cpuHeld()) {
				return tickCompleted || state.cycleBudgetRemaining !== previousRemaining
					? InstructionStepResult.Advanced
					: InstructionStepResult.Blocked;
			}
		}
		if (runtime.pendingCall !== 'entry') {
			return InstructionStepResult.Blocked;
		}
		const result = runtime.cpuExecution.runInstruction(state);
		if (cpu.getFrameDepth() === 0) {
			runtime.pendingCall = null;
			runtime.frameScheduler.clearQueuedTime();
			this.abandonFrameState();
		}
		return result;
	}
}
